### DISCUSSION TITLE
doctest feature in `pytest-pyvista`

### DISCUSSION THREAD

**Discussion by tkoyama010** [pyvista/pyvista#5391] [Ideas]
Is it possible to add doctest functionality to `pytest-pyvista`? With the way it works now, we are not aware if the output image of docstring changes depending on the PR.

**Comment by MatthewFlamm** [comment]
I like this idea.  This might be possible, but it might be a side effect rather than a true test.  For example, the test might fail because the `pl.show()` command prints something when the test fails rather than the error being raised itself.  This isn't necessarily a problem, but should be thought through.

The other enhancement required is better handling of nested file structures with the image cache.  Right now, all images are saved into one folder based on test name.  This works for the use case where the majority of tests with image comparisons are in one file.  We will need a better system for docstrings where the number of images will explode.

Another consideration is the space required in the git repo for all of these images.

**Reply by tkoyama010** [reply]
Thank you for your comment. Taking the hash value of the image would save space I thought. However, I realized that using a hash that checks for identity would not make sense because there is a difference check.

**Reply by tkoyama010** [reply]
I found a library called [imagehash](https://github.com/JohannesBuchner/imagehash) that seems to be able to get the hamming distance, which indicates the similarity of images, but we have not been able to investigate how useful this would be in our case.

**Reply by tkoyama010** [reply]
I am investigating the above. If it is possible to use it, we could write the following in the docstring and have the hash values compared.
```python
        >>> pl.show()
        ffd7918181c9ffff
```

**Reply by MatthewFlamm** [reply]
Storing a hash reduces storage, but it will be more brittle than how we do it now.  In my opinion, it will be easier to test hashing with the current plotting tests.  Then we can know whether it is viable for our tests.

**Reply by MatthewFlamm** [reply]
Although if the hashing could be tuned, we can always choose something that has a low false positive (falsely failing tests) rate to make it only detect very very different images.  This might be a better use for the docstrings then the tests themselves.  So maybe ignore my above comment.

**Reply by banesullivan** [reply]
I think a perceptual hash is the way to go for this, and if I understand correctly, they can allow for a tolerance like we use now. 

@tkoyama010, nice find with [imagehash](https://github.com/JohannesBuchner/imagehash) this looks perfect.

I want to advocate that we do not switch the default for our unit tests to only use hashes because there is immense value in storing the baseline image itself in version control. I've used this several times to go back and see what the test is supposed to look like and what it looks like after changes either in PyVista, OSMesa, or VTK.

I think it would be great to have a `repr` method somewhere for producing these perceptual hashes, which could be used for doctest verification like:

```py
>>> pl.show()
<PyVistaPerceptualHash(ffd7918181c9ffff)>
```

I think the route forward here would be to change our image comparison mechanisms to use perceptual hashes entirely. We'll still save a baseline image in unit tests, but upon comparison, we'll compute the hash of both the baseline and the test image to find the "hamming distance" or other relevant metric for the pass/fail analysis.

Then we could add some sort of hook in `pl.show()` or a repr to the plotter that will generate this hash for doctest comparisons. the problem is that, AFAIK, doctest looks for exact matches to the output... we may need to investigate if it is possible for doctest to leverage an `__equals__` operator on a custom class that lets us handle the comparison internally.

**Reply by MatthewFlamm** [reply]
> I think the route forward here would be to change our image comparison mechanisms to use perceptual hashes entirely.

I think we could easily provide a config option for pytest for `pytest_pyvista` to choose the method.  This way we could opt to run both in parallel to test stability.  

> Then we could add some sort of hook in pl.show() or a repr to the plotter that will generate this hash for doctest comparisons. the problem is that, AFAIK, doctest looks for exact matches to the output... we may need to investigate if it is possible for doctest to leverage an __equals__ operator on a custom class that lets us handle the comparison internally.

My understanding is that pytest uses a character by character comparison unless it detects a number, then it sometimes uses `pytest.approx` comparison depending on usage.  We could store the hashes in a json or similar and then print an error message when the hashes don't match.  This would enable us to add hashes incrementally too.
